service/history/workflow/state_machine_timers.go

```diff
@@ -55,17 +55,28 @@ func AddNextStateMachineTimerTask(ms MutableState) {
 
 // TrackStateMachineTimer tracks a timer task in the mutable state's StateMachineTimers slice sorted and grouped by
 // deadline.
+// Only a single task for a given type can be tracked for a given machine. If a task of the same type is already
+// tracked, it will be overridden.
 func TrackStateMachineTimer(ms MutableState, deadline time.Time, taskInfo *persistencespb.StateMachineTaskInfo) {
 	execInfo := ms.GetExecutionInfo()
 	group := &persistencespb.StateMachineTimerGroup{
 		Deadline: timestamppb.New(deadline),
 		Infos:    []*persistencespb.StateMachineTaskInfo{taskInfo},
 	}
-	idx, found := slices.BinarySearchFunc(execInfo.StateMachineTimers, group, func(a, b *persistencespb.StateMachineTimerGroup) int {
+	idx, groupFound := slices.BinarySearchFunc(execInfo.StateMachineTimers, group, func(a, b *persistencespb.StateMachineTimerGroup) int {
 		return a.Deadline.AsTime().Compare(b.Deadline.AsTime())
 	})
-	if found {
-		execInfo.StateMachineTimers[idx].Infos = append(execInfo.StateMachineTimers[idx].Infos, taskInfo)
+	if groupFound {
+		groupIdx := slices.IndexFunc(execInfo.StateMachineTimers[idx].Infos, func(info *persistencespb.StateMachineTaskInfo) bool {
+			return info.GetType() == taskInfo.GetType() && slices.EqualFunc(info.GetRef().GetPath(), taskInfo.GetRef().GetPath(), func(a, b *persistencespb.StateMachineKey) bool {
```

Comment:
I feel we also need to check taskInfo.Data and ref.InitialVersionedTransition (to support deletion)

Reply:
We don't use `Data` today and we don't expect to revive state machines in the current set of use cases. If we planned to extend the usage of the HSM framework, I would bother, but I don't think it's needed.

Reply:
synced offline, we don't have such use cases so those two checks will essentially always pass.